### DISCUSSION TITLE
Some additional questions on the results generated by `getBCsymmetry.py` tool.

### DISCUSSION THREAD

**Discussion by hongyi-zhao** [goodluck1982/SpaceGroupIrep#20] [General]
This is a cont'd discussion based on the results generated by `getBCsymmetry.py` tool [here](https://github.com/goodluck1982/SpaceGroupIrep/issues/1#issuecomment-1019803052), as denoted by the following content snippet in it: 

```
# 4   : C2                              C2                              C2                              C2
  -1   0   0    0.00002000        -1   0   0    2.00000000        -1   0   0    2.00000000        -1   0   0    2.00000000
   0  -1   0    0.00000000         0  -1   0    0.00000000         0  -1   0    0.00000000         0  -1   0    0.00000000
   0   0   1    0.50000000         0   0   1    0.50000000         0   0   1    0.50000000         0   0   1    0.50000000
[...]
# 7   : md3                             md3                             md3                             md3
   0   1   0    0.00001000         0   1   0    1.00000000         0   1   0    1.00000000         0   1   0    1.00000000
   1   0   0    0.99999000         1   0   0    0.00000000         1   0   0    0.00000000         1   0   0    0.00000000
   0   0   1    0.50000000         0   0   1    0.50000000         0   0   1    0.50000000         0   0   1    0.50000000
# 8   : mv2                             mv2                             mv2                             mv2
   1   0   0    0.00000000         1   0   0    0.00000000         1   0   0    0.00000000         1   0   0    0.00000000
   1  -1   0    0.99999000         1  -1   0    0.00000000         1  -1   0    0.00000000         1  -1   0    0.00000000
   0   0   1    0.00000000         0   0   1    0.00000000         0   0   1    0.00000000         0   0   1    0.00000000
# 9   : md1                             md1                             md1                             md1
   1  -1   0    0.00000000         1  -1   0    0.00000000         1  -1   0    0.00000000         1  -1   0    0.00000000
   0  -1   0    0.00000000         0  -1   0    0.00000000         0  -1   0    0.00000000         0  -1   0    0.00000000
   0   0   1    0.50000000         0   0   1    0.50000000         0   0   1    0.50000000         0   0   1    0.50000000
# 10  : mv3                             mv3                             mv3                             mv3
   0  -1   0    0.00001000         0  -1   0    1.00000000         0  -1   0    1.00000000         0  -1   0    1.00000000
  -1   0   0    0.00001000        -1   0   0    1.00000000        -1   0   0    1.00000000        -1   0   0    1.00000000
   0   0   1    0.00000000         0   0   1    0.00000000         0   0   1    0.00000000         0   0   1    0.00000000
# 11  : md2                             md2                             md2                             md2
  -1   0   0    0.00002000        -1   0   0    2.00000000        -1   0   0    2.00000000        -1   0   0    2.00000000
  -1   1   0    0.00001000        -1   1   0    1.00000000        -1   1   0    1.00000000        -1   1   0    1.00000000
   0   0   1    0.50000000         0   0   1    0.50000000         0   0   1    0.50000000         0   0   1    0.50000000
# 12  : mv1                             mv1                             mv1                             mv1
  -1   1   0    0.00002000        -1   1   0    2.00000000        -1   1   0    2.00000000        -1   1   0    2.00000000
   0   1   0    0.00000000         0   1   0    0.00000000         0   1   0    0.00000000         0   1   0    0.00000000
   0   0   1    0.00000000         0   0   1    0.00000000         0   0   1    0.00000000         0   0   1    0.00000000
```

I've the following questions on the above result:

1. Why isn't `C2` written as `C2+` consistently in the context used by this tool?
2. In the last column, which denotes translation, why not just represent them in the range "0-1"? More specifically, the `2.00000000`, IMHO, is equivalent to `0.00000000`.
3. What are the concrete meanings of `md1`, `md2`, `md3`, `mv1`, `mv2`, and `mv3` respectively as used above? I've tried to look through [the BC book](https://www.physik.uzh.ch/~mchrzasz/For_goska/groups_solids.pdf), but still failed to figure out their implications.

**Comment by hongyi-zhao** [comment]
1. As for question 1, I understand: `C2` is a 180-degree rotation, equivalent to `C2+` or `C2-`, in other words, it is essentially an inversion. Therefore, there is needless to use `+/-` symbols in this situation.

2. As for question 2, I'm not sure if I understand, but have the following impression as given by [gemmi](https://gemmi.readthedocs.io/en/latest/symmetry.html):

```
Wrapping applies modulo 1 to the translational part. 
Which is usually desirable for crystallographic symmetry. 
But the triplets and matrices may represent also, for example, 
generation of a biological assembly. For this reason, x,y+1,z
is not automatically reduced to the identity. 
But when operations are combined, they are assumed to be symmetry
operations and the result is wrapped to [0,1):

op
<gemmi.Op("-y,x-y,z+1/3")>

op * op
<gemmi.Op("-x+y,-x,z+2/3")>

op * op * op  # without wrapping we'd have z+1
<gemmi.Op("x,y,z")>
```
But I really don't know if `getBCsymmetry.py` tool is out of the same considerations.

3. As for question 3, based on the source code definition [here](https://github.com/goodluck1982/SpaceGroupIrep/blob/dbc00ce93e4251ae7dd9029504496da36e5d4edd/getBCsymmetry.py#L101-L106), I figured out the following meanings corresponding to them respectively:

![image](https://user-images.githubusercontent.com/11155854/150940959-0fedc698-3634-47e2-825a-53b1b1ee6ba5.png)



Regards,
HZ

**Comment by goodluck1982** [comment]
> * Why isn't `C2` written as `C2+` consistently in the context used by this tool?

We use the name defined in the BC book.  For  (b) Hexagonal, the name is C2, not C2+.
![image](https://user-images.githubusercontent.com/48718961/150774241-1a0896fa-0d27-498e-a580-d337e0ab316e.png)


> * In the last column, which denotes translation, why not just represent them in the range "0-1"? More specifically, the `2.00000000`, IMHO, is equivalent to `0.00000000`.

In fact, the four columns of the output of getBCsymmetry.py correspond to the four cells below (see the Sec.8.2 of the SpaceGroupIrep paper for detail):
![image](https://user-images.githubusercontent.com/48718961/150964128-1ef2228d-eb23-45b6-b4ac-dc64cc7eb633.png)
The first column is the SG (space group) operations determined by spglib from the input cell, and other columns are the corresponding SG operations **converted** from the first column using equations (19)--(30) in the SpaceGroupIrep paper.

From the viewpoint of SG, {R|000} and {R|001} are two different elements, and they can have different representation matrixes. Therefore it's necessary to reserve the real translation, not the wrapped ones.


> * What are the concrete meanings of `md1`, `md2`, `md3`, `mv1`, `mv2`, and `mv3` respectively as used above? I've tried to look through [the BC book](https://www.physik.uzh.ch/~mchrzasz/For_goska/groups_solids.pdf), but still failed to figure out their implications.

Here m means mirror.  md1 is just $\sigma_{d1}$  in the mathematica output. Because python output can not display ![image](https://user-images.githubusercontent.com/48718961/150965431-f9a27dd1-2231-4dbc-8380-1584d870efcc.png)
 in text form, we use md1 to replace it.

**Comment by hongyi-zhao** [comment]
> The first column is the SG (space group) operations determined by spglib from the input cell, and other columns are the corresponding SG operations **converted** from the first column using equations (19)--(30) in the SpaceGroupIrep paper.

It's a rather complicated procedure, at least viewed from my current knowledge level. BTW, the [arxiv](http://arxiv.org/abs/2012.08871) version of this paper is very long. Why is the published one so short? 

> From the viewpoint of SG, {R|000} and {R|001} are two different elements, and they can have different representation matrixes. Therefore it's necessary to reserve the real translation, not the wrapped ones.

If we consider the ultimate physical meaning of crystallography, they are really the same thing. So, if we finally use these results to analyze the corresponding physical properties and phenomena, the result of the appropriate modulo operation must be used.

Based on my simple and limited observation of the four cells generated by `getBCsymmetry.py` from [`MnO.cif`](https://github.com/goodluck1982/SpaceGroupIrep/issues/1#issuecomment-1019803052), the only differences lie in the translation parts (`v` in Seitz notation `{R|v}`), varied by integer times lattice basis vectors. But I'm not sure if this is the general rule, when using any crystal structure, existing among the four set of cells generated by `getBCsymmetry.py`. 

> Here m means mirror. md1 is just $\sigma_{d1}$ in the mathematica output.

This is consistent with the general rules to naming [the mirror operation](http://newton.ex.ac.uk/research/qsystems/people/goss/symmetry/CharacterTables.html):

![image](https://user-images.githubusercontent.com/11155854/150979296-1e0c052b-9a04-4f70-af6a-7ff4c55503f2.png)

But it seems that for a hexagonal system, the mirror symmetries are more rich and complex than the ones described in the above table, as you have defined in the `getBCsymmetry.py`: `$\sigma_{v1}$`, `$\sigma_{v2}$`, `$\sigma_{v3}$`, `$\sigma_{d1}$`, `$\sigma_{d2}$`, and `$\sigma_{d3}$`. But I still can't figure out the numbering rules for subscript in the above naming.  

In particular, I am still unclear about the symmetry represented by `M̃z = { Mz|0,0,1/2 }` in [this paper](https://doi.org/10.1016/j.physleta.2020.126494):

![image](https://user-images.githubusercontent.com/11155854/151098682-f81577c3-6179-48a8-ab9e-82a9cd0519d3.png)

More specifically, your `getBCsymmetry.py` tool will give the following results for the mirror symmetry that this material has:

```
$ python getBCsymmetry.py -c ~/Downloads/caj/spin-gapless/ICSD_CollCode262928.cif -p 0.0001
[...]
# 7   : md3                             md3                             md3                             md3  
   0   1   0    0.00001000         0   1   0    1.00000000         0   1   0    1.00000000         0   1   0    1.00000000 
   1   0   0    0.99999000         1   0   0    0.00000000         1   0   0    0.00000000         1   0   0    0.00000000 
   0   0   1    0.50000000         0   0   1    0.50000000         0   0   1    0.50000000         0   0   1    0.50000000 
# 8   : mv2                             mv2                             mv2                             mv2  
   1   0   0    0.00000000         1   0   0    0.00000000         1   0   0    0.00000000         1   0   0    0.00000000 
   1  -1   0    0.99999000         1  -1   0    0.00000000         1  -1   0    0.00000000         1  -1   0    0.00000000 
   0   0   1    0.00000000         0   0   1    0.00000000         0   0   1    0.00000000         0   0   1    0.00000000 
# 9   : md1                             md1                             md1                             md1  
   1  -1   0    0.00000000         1  -1   0    0.00000000         1  -1   0    0.00000000         1  -1   0    0.00000000 
   0  -1   0    0.00000000         0  -1   0    0.00000000         0  -1   0    0.00000000         0  -1   0    0.00000000 
   0   0   1    0.50000000         0   0   1    0.50000000         0   0   1    0.50000000         0   0   1    0.50000000 
# 10  : mv3                             mv3                             mv3                             mv3  
   0  -1   0    0.00001000         0  -1   0    1.00000000         0  -1   0    1.00000000         0  -1   0    1.00000000 
  -1   0   0    0.00001000        -1   0   0    1.00000000        -1   0   0    1.00000000        -1   0   0    1.00000000 
   0   0   1    0.00000000         0   0   1    0.00000000         0   0   1    0.00000000         0   0   1    0.00000000 
# 11  : md2                             md2                             md2                             md2  
  -1   0   0    0.00002000        -1   0   0    2.00000000        -1   0   0    2.00000000        -1   0   0    2.00000000 
  -1   1   0    0.00001000        -1   1   0    1.00000000        -1   1   0    1.00000000        -1   1   0    1.00000000 
   0   0   1    0.50000000         0   0   1    0.50000000         0   0   1    0.50000000         0   0   1    0.50000000 
# 12  : mv1                             mv1                             mv1                             mv1  
  -1   1   0    0.00002000        -1   1   0    2.00000000        -1   1   0    2.00000000        -1   1   0    2.00000000 
   0   1   0    0.00000000         0   1   0    0.00000000         0   1   0    0.00000000         0   1   0    0.00000000 
   0   0   1    0.00000000         0   0   1    0.00000000         0   0   1    0.00000000         0   0   1    0.00000000 
[...]
```
OTOH, the supplementary Material written by [Daniel B. Litvin](https://journals.iucr.org/a/issues/2011/04/00/pz5089/pz5089sup1.pdf) gives the following description of Seitz notation for space group 186 (P6_3mc):

![image](https://user-images.githubusercontent.com/11155854/151099234-ff527359-85ff-494d-8b88-b4ba6d64c187.png)

So, I would like to know which operators in the output of `getBCsymmetry.py` correspond to the Seitz notation `M̃z = { Mz|0,0,1/2 }` used in the above [paper](https://doi.org/10.1016/j.physleta.2020.126494). And what is the correspondence of the results given by `getBCsymmetry.py` and those by [Daniel B. Litvin](https://journals.iucr.org/a/issues/2011/04/00/pz5089/pz5089sup1.pdf).

I also find that the mirror symmetry operators of SG 186 given by `getBCsymmetry.py` don't, in terms of representation,  correspond one-to-one with the ones given by [`cctbx`](http://cci.lbl.gov/cctbx/explore_symmetry.html), as shown below.

The operators used by `getBCsymmetry.py`:

![image](https://user-images.githubusercontent.com/11155854/150940959-0fedc698-3634-47e2-825a-53b1b1ee6ba5.png)

The operators given by [`cctbx`](http://cci.lbl.gov/cctbx/explore_symmetry.html):

![image](https://user-images.githubusercontent.com/11155854/151101947-95eec36a-4be2-47dd-90d5-85ceac6ec70c.png)

**Side remarks**:

1. According to the notes [here](https://github.com/goodluck1982/SpaceGroupIrep#capabilities-of-spacegroupirep) and [here](https://github.com/goodluck1982/SpaceGroupIrep#files), the `getBCsymmetry.py` script just provides a subset of the complete capabilities of `SpaceGroupIrep` package for user convenience.

2. OTOH, Mathematica is powerful enough to be omnipotent on almost any problem in natural science, but it's not free and the learning curve is relatively steep, which will limit some users from using the package.

**Comment by goodluck1982** [comment]
> > The first column is the SG (space group) operations determined by spglib from the input cell, and other columns are the corresponding SG operations **converted** from the first column using equations (19)--(30) in the SpaceGroupIrep paper.
> 
> It's a rather complicated procedure, at least viewed from my current knowledge level. BTW, the [arxiv](http://arxiv.org/abs/2012.08871) version of this paper is very long. Why is the published one so short?

Yes, it's somewhat complicated.  The original purpose of getBCsymmetry.py is for debug, and it's in fact a internal tool for myself. I choose to make it public only because it can convert a non-BC cell to a BC cell and generate corresponding vasp POSCAR file. If you want to know all the details of the output information of getBCsymmetry.py, you have to understand the Sec8.2 of the SpaceGroupIrep paper completely. 

As for the paper length. The arxiv version contains the supplementary materail of hundreds of pages, while in the published version the supplementary materail is supplied alone.
 
> > From the viewpoint of SG, {R|000} and {R|001} are two different elements, and they can have different representation matrixes. Therefore it's necessary to reserve the real translation, not the wrapped ones.
> 
> If we consider the ultimate physical meaning of crystallography, they are really the same thing. So, if we finally use these results to analyze the corresponding physical properties and phenomena, the result of the appropriate modulo operation must be used.
> 
> Based on my simple and limited observation of the four cells generated by `getBCsymmetry.py` from [`MnO.cif`](https://github.com/goodluck1982/SpaceGroupIrep/issues/1#issuecomment-1019803052), the only differences lie in the translation parts (`v` in Seitz notation `{R|v}`), varied by integer times lattice basis vectors. But I'm not sure if this is the general rule, when using any crystal structure, existing among the four set of cells generated by `getBCsymmetry.py`.
> 

This involves the cell conversion procedure. If you want to understand it, you have to first understand the Sec8.2 of the SpaceGroupIrep paper completely.  The differenncs among the four sets of data depend on the input cell case by case. As for the translations, of course you can only care about the wrapped ones. {R|000} and {R|001} may be the same to you, but they are different to me for  debug purpose. 

So, you can just omit the "redundant integer translations" as you understand.



> > Here m means mirror. md1 is just $\sigma_{d1}$ in the mathematica output.
> 
> This is consistent with the general rules to naming [the mirror operation](http://newton.ex.ac.uk/research/qsystems/people/goss/symmetry/CharacterTables.html):
>  
> But it seems that for a hexagonal system, the mirror symmetries are more rich and complex than the ones described in the above table, as you have defined in the `getBCsymmetry.py`: `$\sigma_{v1}$`, `$\sigma_{v2}$`, `$\sigma_{v3}$`, `$\sigma_{d1}$`, `$\sigma_{d2}$`, and `$\sigma_{d3}$`. But I still can't figure out the numbering rules for subscript in the above naming.

It should be kept in mind that the rotation names are **NOT defined by me, but by the BC book** [“The mathematical theory of symmetry in solids” by C. J. Bradley & A. P. Cracknell]

> 
> In particular, I am still unclear about the symmetry represented by `M̃z = { Mz|0,0,1/2 }` in [this paper](https://doi.org/10.1016/j.physleta.2020.126494):
> 
> More specifically, your `getBCsymmetry.py` tool will give the following results for the mirror symmetry that this material has:
> 
> OTOH, the supplementary Material written by [Daniel B. Litvin](https://journals.iucr.org/a/issues/2011/04/00/pz5089/pz5089sup1.pdf) gives the following description of Seitz notation for space group 186 (P6_3mc):
>  
> So, I would like to know which operators in the output of `getBCsymmetry.py` correspond to the Seitz notation `M̃z = { Mz|0,0,1/2 }` used in the above [paper](https://doi.org/10.1016/j.physleta.2020.126494). And what is the correspondence of the results given by `getBCsymmetry.py` and those by [Daniel B. Litvin](https://journals.iucr.org/a/issues/2011/04/00/pz5089/pz5089sup1.pdf).
> 
> I also find that the mirror symmetry operators of SG 186 given by `getBCsymmetry.py` don't, in terms of representation, correspond one-to-one with the ones given by [`cctbx`](http://cci.lbl.gov/cctbx/explore_symmetry.html), as shown below.
> 
> The operators used by `getBCsymmetry.py`:
> 
> The operators given by [`cctbx`](http://cci.lbl.gov/cctbx/explore_symmetry.html):
> 

You cite several places (paper or webpage). You should know that **the naming and cell convention can be different place by place**.   In principle, everyone can define his/her own convertion.  And we use the convention of the BC book. To make a correspondence  between two conventions A and B, one should first knows complete information of both A and B, including the cell definition, the naming of rotations, and so on. Some papers do not mention clearly the convention they use, then one have to guess and try.  

Back to your problem:  the correspondence between the BC convention and a convention in a paper.
You should first make clear the BC convention. The best method is to read the BC book, such as the following parts:
![image](https://user-images.githubusercontent.com/48718961/151115991-b4081ad7-136d-451f-8b8c-96cd2559663c.png)
![image](https://user-images.githubusercontent.com/48718961/151116329-0b644c34-1dbd-4e86-a9dd-438f39047aa7.png)
Then, you should make clear the convention used in the paper.
After these two procedures, you can understand their relation.
 
> **Side remarks**:
> 
> 1. According to the notes [here](https://github.com/goodluck1982/SpaceGroupIrep#capabilities-of-spacegroupirep) and [here](https://github.com/goodluck1982/SpaceGroupIrep#files), the `getBCsymmetry.py` script just provides a subset of the complete capabilities of `SpaceGroupIrep` package for user convenience.

 `getBCsymmetry.py`  only provides the cell conversion procedure and the details are not necessarily cared about for the users who only want to obtain a BC cell from any input cell.

> 2. OTOH, Mathematica is powerful enough to be omnipotent on almost any problem in natural science, but it's not free and the learning curve is relatively steep, which will limit some users from using the package.

I understand this. However at present we have no more manpower and time to develop a version using other language, such as python.

**Comment by goodluck1982** [comment]
In your problem, the {Mz|001/2} should be one of the three operations in the red box
![Snipaste_2022-01-26_15-05-34](https://user-images.githubusercontent.com/48718961/151118537-adf61ecb-6f83-4b80-89ca-08f1514c5c7b.jpg)
However, I don't know which one is. Because I don't know the explicit definition of Mz in the paper.

**Comment by hongyi-zhao** [comment]
> To make a correspondence between two conventions A and B, one should first knows complete information of both A and B, including the cell definition, the naming of rotations, and so on.

> Then, you should make clear the convention used in the paper.
> After these two procedures, you can understand their relation.

These heartfelt words are truth and philosophy about learning methods.

> In your problem, the {Mz|001/2} should be one of the three operations in the red box ![Snipaste_2022-01-26_15-05-34](https://user-images.githubusercontent.com/48718961/151118537-adf61ecb-6f83-4b80-89ca-08f1514c5c7b.jpg) 

That's what I think, because among the six mirror symmetry operators of SG 186, only the above three Seitz notations have the translation part `0,0,1/2`, which is consistent with the description in [the paper](https://doi.org/10.1016/j.physleta.2020.126494).

> However, I don't know which one is. Because I don't know the explicit definition of Mz in the paper.

The author does not give any definition of `M_z`, but only mentions the existence of this symmetry. Therefore, my understanding is the same as yours: It is used to represent any or all of the three kinds of symmetries, as indicated by you. 

BTW, if only one of these symmetries, and not all of them, is responsible for the subsequent analysis of specific physical phenomena and results, then the corresponding exact description must be given. This is the trouble that the inaccuracy of the representation used in [the paper](https://doi.org/10.1016/j.physleta.2020.126494) may cause.

**Comment by goodluck1982** [comment]
Maybe which mirror is Mz is not important for the authors of the paper or does not affect the results of the paper.
But we can not think that sigma_d1,d2,d3 are the same. At least their representation matrixes are different:
![image](https://user-images.githubusercontent.com/48718961/151122245-f08333df-d768-41c8-bc7f-b00415f0f868.png)

**Comment by hongyi-zhao** [comment]
What's your OS, and what's the screen capture tool used by you? The above screenshot is rather blurry.

P.S. I use various *nix based distros for about 15 years. The current desktop operating system is Ubuntu 20.04.3 LTS and I use [`shutter`](https://github.com/shutter-project/shutter) or [`ksnip`](https://github.com/ksnip/ksnip) as the screen capture tools.

**Comment by goodluck1982** [comment]
****

> What's your OS, and what's the screen capture tool used by you? The above screenshot is rather blurry.

win 10  &  snipaste
Because I zoomed out the screen to capture the whole picture.
